### ISSUE TITLE
Reducer `axis=None` cases should handle `initial` (pass to NumPy) and raise exceptions for empty arrays

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#1402]
### Version of Awkward Array

1.8.0

### Description and code to reproduce

This might be intentional (I doubt it), but I found it surprising. 
```pycon
>>> ak.max(np.zeros(0, dtype=np.int64), mask_identity=False) # None
>>> ak.max(np.zeros(0, dtype=np.int64), axis=-1, mask_identity=False)
-9223372036854775808
```

I'm fairly confident that this is just an artefact of the special casing for `axis=None`.

There is a separate bug related to this special casing:
```pycon
>>> ak.max(np.zeros(1, dtype=np.int64), initial=1_000_000, mask_identity=False)
0
>>> ak.max(np.zeros(1, dtype=np.int64), axis=-1, initial=1_000_000, mask_identity=False)
1000000
```
We don't handle the `initial` case either.

One solution might be to use to the existing reduction logic for `axis=0` within the completely flattened array parts. That way, all of the important code takes the same code path, and we'd just have a top-level null-coalescing max to build the final solution.

**Comment by jpivarski** [comment]
Without `mask_identity`, empty arrays or lists result in the "identity" of the reducer's semigroup operation. For maximization, the identity is minus infinity. For integer types, which don't have infinite values, the identity is the smallest possible value.

```python
np.iinfo(np.int64).min   →   -9223372036854775808
```

Try it with `np.uint64`; it should be `0`. So, strange as it may seem, it's the `axis=None` case that's wrong here: it should not be returning None. It should be returning `-9223372036854775808`.

Same for the handling of `initial`. It was added because it's an argument that NumPy defines.

You're right that the special-casing is responsible for this bug. However, I don't see how we can remove the special-casing. Maybe the whole array can be put inside of a length-1 RegularArray and the reducer can be applied to it, but that wouldn't handle RecordArrays the way that our chosen behavior for `axis=None` does—flattening all the fields of the records. For that, you need `completely_flatten`, which is completely distinct from the `reduce_next` formalism. It would be easier to apply this fix to the special cases.

**Comment by agoose77** [comment]
Ah, I realise that I didn't put the issue title in the body text

> So, strange as it may seem, it's the axis=None case that's wrong here

I understood the significance of `-9223372036854775808` (and indeed, I wanted an integer result when I stumbled into this bug),  and this identity for `max` makes sense. The bug(s) that I was trying to report was the `axis=None` case ignores both `identity` and `mask_identity`.

As discussed, this happens simply because we don't pass these options to the `nplike` special case block.

The solution that I was proposing was to take the flat array "parts" that are returned by `completely_flatten` and invoke the Awkward reduction mechanism to reduce them, followed by a null-aware outer reducer. I.E., replace the `nplike.min` call with `layout.min`.

**Comment by jpivarski** [comment]
The outer reducer would have to be a special path, regardless. This would be a better thing to fix with minimal changes, I think. I don't see a way to remove its specialness

**Comment by agoose77** [comment]
Yes, we need a special path (unless we implement a partitioned array again, though I assume that just concatenated the memory anyway, which we could do if it were actually desirable!).

I was going to propose a two-stage reduction, which would have simplified the code, but this would only work for reducers that form a group (`add`, `prod`).[^footnote] This doesn't hold for all reducers, e.g. `count`, which maps _to_ a different set (the counts).

[^footnote]: your use of the group terminology was a useful reminder that it's a good concise language here

**Comment by agoose77** [comment]
Additionally, for when we fix this issue, I think the logic in `argmax` is not quite right at the moment. Here we store the best `argmax` result within each sublist. This will not correctly index into the ravelled array if any part besides the first one wins:
https://github.com/scikit-hep/awkward-1.0/blob/278dccee86046f7b92505c6f725ff8574235ecf6/src/awkward/_v2/operations/reducers/ak_argmax.py#L84

We just need to add 
```python
if best_index is None:
    best_index = out
    best_value = tmp[out]
elif tmp[out] > best_value:
    best_index = best_index + out
    best_value = tmp[out]
```

Reproducer:
```python
import awkward as ak
import numpy as np


array = ak._v2.zip({
    "this": np.arange(10),
    "that": np.arange(10) + 10
})

print(ak._v2.argmax(array, axis=None, flatten_records=True))
```

The same bug is in `argmin`

**Comment by jpivarski** [comment]
Avoiding concatenation was intentional: it's a performance thing, but it's a big one. Taking a set of buffers as-is, reducing over each one of them separately, and then reducing those results, is much better than concatenating them all into a contiguous memory buffer and then reducing that once. (I think even if we had PartitionedArray, it would still be a separate code-path from that, but anyway, it's moot: there's no PartitionedArray.)

I just checked NumPy and the argmin/argmax index is the _first_ index when the extreme value is repeated, and that's what happens without records, but not with records (i.e. when `completely_flatten` returns more than one buffer). There's some policy decision about what argmin/argmax should even mean when you have to linearize the fields into one sequence, but the simplest interpretation (the one I would assume) is that they get flattened into depth-first order.

So

```python
>>> array = ak._v2.zip({
...     "this": np.arange(10),
...     "that": np.arange(10) + 10
... })
>>> array.show()
[{this: 0, that: 10},
 {this: 1, that: 11},
 {this: 2, that: 12},
 {this: 3, that: 13},
 {this: 4, that: 14},
 {this: 5, that: 15},
 {this: 6, that: 16},
 {this: 7, that: 17},
 {this: 8, that: 18},
 {this: 9, that: 19}]
>>> ak._v2.argmax(array, axis=None, flatten_records=True)
9
```

shouldn't be `9`, it should be `19`, and

```python
>>> array = ak._v2.zip({
...     "this": np.arange(10) + 10,
...     "that": np.arange(10)
... })
>>> array.show()
[{this: 10, that: 0},
 {this: 11, that: 1},
 {this: 12, that: 2},
 {this: 13, that: 3},
 {this: 14, that: 4},
 {this: 15, that: 5},
 {this: 16, that: 6},
 {this: 17, that: 7},
 {this: 18, that: 8},
 {this: 19, that: 9}]
>>> ak._v2.argmin(array, axis=None, flatten_records=True)
0
```

shouldn't be `0`, it should be `10`.

**Comment by agoose77** [comment]
> Avoiding concatenation was intentional: it's a performance thing, but it's a big one. 

Absolutely, if concatenation were cheap we'd not need any of this code! Again, to clarify, I'm really saying "which we would have  done if it were actually desirable!"

> There's some policy decision about what argmin/argmax should even mean when you have to linearize the fields into one sequence, but the simplest interpretation (the one I would assume) is that they get flattened into depth-first order.

Right, that's my interpretation too. By "first wins" I was referring to the "part" (partition) not the unique value. In our existing implementation, we store the per-partition index, not the ravelled index.

**Comment by jpivarski** [comment]
I changed the title to make a firm decision on this. Although it was designed to return `None` on empty arrays (note the explicit type-tracer code for that), this is not what NumPy does. As annoying as I consider NumPy's behavior, we should reproduce it:

```python
>>> np.max([])
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in amax
  File "/home/jpivarski/mambaforge/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 2791, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "/home/jpivarski/mambaforge/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 86, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```

Similarly, `initial` should be passed to the NumPy function, since NumPy (nplike) handles that already.

This one should be pretty quick, though I've just run out of time in my Day of Fixing Bugs.

**Comment by agoose77** [comment]
Closed by #2020
